R/run.R

```diff
@@ -0,0 +1,106 @@
+#' run
+#'
+#' Run one or multiple climate models over one or multiple emissions scenarios.
+#'
+#' Some simple climate models need certain information to run. Use the setup function to supply this information
+#' before attempting to use run.
+#'
+#' @param climateModelsConfigs Named list where the names are climate models and the corresponding list member is a
+#' configuration or list of configurations to use for this model. Use null to denote the default configuration with
+#' no changes. Climate model names supported: CiceroSCM, FaIR, and MAGICC7. The supported configuration settings
+#' depend on the used climate model, please refer to their respective documentation.
+#' @param scenarios DataFrame containing one or multiple emissions scenarios to simulate. The data shoud be in
+#' IAMC format, i.e. a wide data frame with the years as columns. Index columns that are additionally required are
+#' "model", "scenario", "region", "variable", and "unit". The "variable" must also follow the IAMC format naming
+#' conventions, e.g. "Emissions|CO2" for the 'total CO2 emissions (not including CCS)'.
+#' @param outputVariables A list of variables to include in the output. Optional, default: list("Surface Temperature")
+#' @param outConfig Named list where the names are climate models and the corresponding list member is a list of
+#' configuration values to include in the output in the metadata. Optional, default: don't include input variables in
+#' the output metadata.
+#' @param returnRaw Boolean to control the return type. By default (returnRaw = FALSE), run() returns a named list
+#' result, where result$df is a data frame which contains the result in the same format as the input scenarios, and
+#' result$metadata is a named list of metadata generated during the run. If returnRaw is TRUE, run() returns a python
+#' scmdata.ScmRun object.
+#' @author Mika Pflüger
+#' @importFrom reticulate import py_to_r
+#' @examples
+#' \dontrun{
+#' # create very minimal emissions scenario.
+#' df <- data.frame(
+#'  model = c("rand", "rand"),
+#'  scenario = c("weirdEMI", "weirdEMI"),
+#'  region = c("World", "World"),
+#'  variable = c("Emissions|CO2", "Emissions|CH4"),
+#'  unit = c("Mt CO2 / yr", "Mt CH4 / yr"),
+#'  "2015" = c(9., 12.),
+#'  "2020" = c(10., 11.),
+#'  check.names = FALSE)
+#'
+#' # simulate the scenario using MAGICC7 with default settings.
+#' run(climateModelsConfigs = list("MAGICC7" = NULL), scenarios = df)
```

Comment:
Hm… I'm seriously thinking of training myself to always use quotes in named list definitions, so that I never have to think about the question "is this a valid R identifier". Apparently `MAGICC7` is a valid R identifier - but is it more readable to not use quotes everywhere?

Reply:
That's fair, I'd say it comes down to personal preference. I personally find it irritating to have strings to the left of `=`. If something is not a valid R identifier I'd prefer ```list(`MAGICC7` = NULL)``` just to distinguish it from a normal string value,so could also always use backticks in these situations.

Reply:
I guess it is because I learned Python first, and in dictionaries (~named lists) the keys can be any hashable type (so, generally, whatever). Therefore, for me the names / keys are "data", not "code", and the thinking "this is a normal string value" is actually the correct one. I guess in R that is subtly wrong because I can `a$MAGICC7` in this case and there it obviously isn't data, but code. Honestly, it feels so wrong, and produces all of these `check.names = FALSE` problems. But probably it would be more idiomatic in R to embrace that there is no general hashed map data structure…